src/execution_engine.rs

```diff
@@ -74,45 +80,71 @@ impl<TS: TypeSystem> ExecutionEngine<TS> {
         self.globals = vec![Value::uninitialized_reference(); self.num_globals];
     }
 
+    #[inline]
     pub fn call(
         &mut self,
         func: &FunctionRef<TS>,
-        mut args: Vec<TS::Value>,
+        args: impl IntoExactSizeIterator<Item = TS::Value>,
     ) -> Result<TS::Value, FreightError> {
-        if !func.arg_count.valid_arg_count(args.len()) {
+        let vec = BoxSlicePool::request(self.box_pool.clone(), func.stack_size);
+        let mut iter = args.into_exact_size_iter();
+        let arg_count = iter.len();
+        self.call_internal(func, vec, |_| Ok(iter.next().unwrap()), arg_count)
+    }
+
+    pub(crate) fn call_internal(
+        &mut self,
+        func: &FunctionRef<TS>,
+        mut stack: Stack<TS::Value>,
+        mut args: impl FnMut(&mut ExecutionEngine<TS>) -> Result<TS::Value, FreightError>,
+        arg_count: usize,
+    ) -> Result<TS::Value, FreightError> {
+        if !func.arg_count.valid_arg_count(arg_count) {
             return Err(FreightError::IncorrectArgumentCount {
                 expected_min: func.arg_count.min(),
                 expected_max: func.arg_count.max(),
-                actual: args.len(),
+                actual: arg_count,
             });
         }
-
-        while args.len() < func.arg_count.max_capped() {
-            args.push(Value::uninitialized_reference());
+        let mut arg_num = 0;
+        let max = func.arg_count.max_capped().min(arg_count);
+        while arg_num < max {
+            stack[arg_num] = args(self)?;
+            arg_num += 1;
         }
+        stack[arg_num..].fill_with(Value::uninitialized_reference);
 
         #[cfg(feature = "variadic_functions")]
-        if let ArgCount::Variadic { min: _, max } = func.arg_count {
-            let vargs = args.split_off(max);
-            args.push(crate::value::Value::gen_list(vargs));
+        if let ArgCount::Variadic { .. } = func.arg_count {
+            let mut vargs = Vec::new();
+            let mut index = arg_num;
+            while index < arg_count {
+                vargs.push(args(self)?);
+                index += 1;
+            }
```

Comment:
why cant this be a for loop?
```suggestion
            for _ in arg_num..arg_count{
                vargs.push(args(self)?);
            }
```